// @sanity/preview-url-secret/dist/constants.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
const schemaType = "sanity.previewUrlSecret", schemaIdPrefix = "sanity-preview-url-secret", schemaIdSingleton = `${schemaIdPrefix}.share-access`, schemaTypeSingleton = "sanity.previewUrlShareAccess", apiVersion = "2023-11-09", urlSearchParamPreviewSecret = "sanity-preview-secret", urlSearchParamPreviewPathname = "sanity-preview-pathname", urlSearchParamPreviewPerspective = "sanity-preview-perspective", isDev = process.env.NODE_ENV === "development", SECRET_TTL = 60 * 60, fetchSecretQuery = (
  /* groq */
  `*[_type == "${schemaType}" && secret == $secret && dateTime(_updatedAt) > dateTime(now()) - ${SECRET_TTL}][0]{
    _id,
    _updatedAt,
    secret,
    studioUrl,
  }`
), fetchSharedAccessQuery = (
  /* groq */
  `*[_id == "${schemaIdSingleton}" && _type == "${schemaTypeSingleton}"][0].secret`
), fetchSharedAccessSecretQuery = (
  /* groq */
  `*[_id == "${schemaIdSingleton}" && _type == "${schemaTypeSingleton}" && secret == $secret][0]{
  secret,
  studioUrl,
}`
), deleteExpiredSecretsQuery = (
  /* groq */
  `*[_type == "${schemaType}" && dateTime(_updatedAt) <= dateTime(now()) - ${SECRET_TTL}]`
), tag = "sanity.preview-url-secret", perspectiveCookieName = "sanity-preview-perspective";
exports.SECRET_TTL = SECRET_TTL;
exports.apiVersion = apiVersion;
exports.deleteExpiredSecretsQuery = deleteExpiredSecretsQuery;
exports.fetchSecretQuery = fetchSecretQuery;
exports.fetchSharedAccessQuery = fetchSharedAccessQuery;
exports.fetchSharedAccessSecretQuery = fetchSharedAccessSecretQuery;
exports.isDev = isDev;
exports.perspectiveCookieName = perspectiveCookieName;
exports.schemaIdPrefix = schemaIdPrefix;
exports.schemaIdSingleton = schemaIdSingleton;
exports.schemaType = schemaType;
exports.schemaTypeSingleton = schemaTypeSingleton;
exports.tag = tag;
exports.urlSearchParamPreviewPathname = urlSearchParamPreviewPathname;
exports.urlSearchParamPreviewPerspective = urlSearchParamPreviewPerspective;
exports.urlSearchParamPreviewSecret = urlSearchParamPreviewSecret;
