// @sanity/presentation/dist/_chunks-cjs/PostMessageTelemetry.cjs
"use strict";
var reactCompilerRuntime = require("react-compiler-runtime"), react = require("react");
function createNoopLogger() {
  const logger = {
    updateUserProperties() {
    },
    trace,
    log
  };
  function trace(telemetryTrace) {
    return {
      start() {
      },
      log(data) {
      },
      complete() {
      },
      newContext(name) {
        return logger;
      },
      error(error) {
      },
      await: (promise) => promise
    };
  }
  function log(event, data) {
  }
  return logger;
}
const noopLogger = createNoopLogger(), TelemetryLoggerContext = react.createContext(noopLogger);
function useTelemetry() {
  return react.useContext(TelemetryLoggerContext);
}
const PostMessageTelemetry = (props) => {
  const $ = reactCompilerRuntime.c(4), {
    comlink
  } = props, telemetry = useTelemetry();
  let t0, t1;
  return $[0] !== comlink || $[1] !== telemetry ? (t0 = () => comlink.on("visual-editing/telemetry-log", async (message) => {
    const {
      event,
      data
    } = message;
    data ? telemetry.log(event, data) : telemetry.log(event);
  }), t1 = [comlink, telemetry], $[0] = comlink, $[1] = telemetry, $[2] = t0, $[3] = t1) : (t0 = $[2], t1 = $[3]), react.useEffect(t0, t1), null;
};
var PostMessageTelemetry$1 = react.memo(PostMessageTelemetry);
exports.default = PostMessageTelemetry$1;
