// @sanity/vision/lib/_chunks-cjs/index.js
"use strict";
var icons = require("@sanity/icons"), react = require("react"), sanity = require("sanity"), router = require("sanity/router");
const visionLocaleNamespace = "vision", visionUsEnglishLocaleBundle = sanity.defineLocaleResourceBundle({
  locale: "en-US",
  namespace: visionLocaleNamespace,
  resources: () => Promise.resolve().then(function() {
    return require("./resources.js");
  })
});
var __getOwnPropSymbols = Object.getOwnPropertySymbols, __hasOwnProp = Object.prototype.hasOwnProperty, __propIsEnum = Object.prototype.propertyIsEnumerable, __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    __hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0 && (target[prop] = source[prop]);
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source))
      exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop) && (target[prop] = source[prop]);
  return target;
};
const visionTool = sanity.definePlugin((options) => {
  const _a = options || {}, {
    name,
    title,
    icon
  } = _a, config = __objRest(_a, [
    "name",
    "title",
    "icon"
  ]);
  return {
    name: "@sanity/vision",
    tools: [{
      name: name || "vision",
      title: title || "Vision",
      icon: icon || icons.EyeOpenIcon,
      component: react.lazy(() => Promise.resolve().then(function() {
        return require("./SanityVision.js");
      })),
      options: config,
      router: router.route.create("/*"),
      __internalApplicationType: "sanity/vision"
    }],
    i18n: {
      bundles: [visionUsEnglishLocaleBundle]
    }
  };
});
exports.visionLocaleNamespace = visionLocaleNamespace;
exports.visionTool = visionTool;
